rust/protocol/src/error.rs

```diff
@@ -11,82 +15,208 @@ use uuid::Uuid;
     UntrustedIdentity(crate::ProtocolAddress),
 
     /// invalid prekey identifier
+    ///
+    /// Raised if a [PreKeyId] could not be resolved to a [PreKeyRecord] by a [PreKeyStore].
     InvalidPreKeyId,
     /// invalid signed prekey identifier
+    ///
+    /// Raised if a [SignedPreKeyId] could not be resolved to a [SignedPreKeyRecord] by
+    /// a [SignedPreKeyStore].
     InvalidSignedPreKeyId,
 
     /// invalid MAC key length <{0}>
+    ///
+    /// Raised if a MAC key is deserialized from an incorrectly-sized slice.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     InvalidMacKeyLength(usize),
 
     /// missing sender key state for distribution ID {distribution_id}
-    NoSenderKeyState { distribution_id: Uuid },
+    ///
+    /// Raised if a [SenderKeyStore] is unable to locate a [SenderKeyRecord] for a given
+    /// *([ProtocolAddress], [Uuid])* pair in [group_encrypt] or [group_decrypt].
+    NoSenderKeyState {
+        /// Unique identifier for the sender key session.
+        distribution_id: Uuid,
+    },
 
     /// session with {0} not found
+    ///
+    /// Raised if an [IdentityKeyStore] does not contain an entry for a [ProtocolAddress], or
+    /// alternately if a [SessionStore] does not contain a session for a given [ProtocolAddress].
     SessionNotFound(crate::ProtocolAddress),
     /// invalid session: {0}
+    ///
+    /// Raised if a [SessionStore] does not contain a remote identity key to validate.
+    ///
+    /// Similar to [Self::InvalidState].
     InvalidSessionStructure(&'static str),
     /// invalid sender key session with distribution ID {distribution_id}
-    InvalidSenderKeySession { distribution_id: Uuid },
+    ///
+    /// Raised if a [SenderKeyStore] could not load the session for the given ID.
+    ///
+    /// Similar to [Self::NoSenderKeyState].
+    InvalidSenderKeySession {
+        /// Unique identifier for the sender key session.
+        distribution_id: Uuid,
+    },
     /// session for {0} has invalid registration ID {1:X}
+    ///
+    /// Raised if a sealed sender message has a registration id that doesn't map to any
+    /// known session.
     InvalidRegistrationId(crate::ProtocolAddress, u32),
 
     /// message with old counter {0} / {1}
+    ///
+    /// Raised if the same message is decrypted twice so it can be discarded.
     DuplicatedMessage(u32, u32),
     /// invalid {0:?} message: {1}
+    ///
+    /// Raised if a [SignalMessage] could not be decrypted or some field had an unexpected value.
+    ///
+    /// *TODO: what differentiates this from [Self::CiphertextMessageTooShort]?*
     InvalidMessage(crate::CiphertextMessageType, &'static str),
 
     /// error while invoking an ffi callback: {0}
+    ///
+    /// Raised to propagate an error from an FFI callback.
     FfiBindingError(String),
```

Comment:
This is not about propagating errors; it's when the callback itself can't be invoked. Think JNI-level errors. (It does not actually come up for pure Rust users and maybe doesn't belong here.)